src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1215,6 +1224,136 @@ multi_relation_restriction_hook(PlannerInfo *root, RelOptInfo *relOptInfo, Index
+	Datum resultIdDatum = 0;
+	char *resultId = NULL;
+	int64 resultSize = 0;
+	ListCell *typeCell = NULL;
+	bool binaryFormat = false;
+	double rowCost = 0.;
+	double rowSizeEstimate = 0;
+	double rowCountEstimate = 0.;
+	double ioCost = 0.;
+
+	if (rangeTableEntry->rtekind != RTE_FUNCTION ||
+		list_length(rangeTableEntry->functions) != 1)
+	{
+		/* avoid more expensive checks below for non-functions */
+		return;
+	}
+
+	if (!CitusHasBeenLoaded() || !CheckCitusVersion(DEBUG5))
+	{
+		/* read_intermediate_result may not exist */
+		return;
+	}
+
+	if (!ContainsReadIntermediateResultFunction((Node *) rangeTableEntry->functions))
+	{
+		return;
+	}
+
+	rangeTableFunction = (RangeTblFunction *) linitial(rangeTableEntry->functions);
+	funcExpression = (FuncExpr *) rangeTableFunction->funcexpr;
+	resultIdConst = (Const *) linitial(funcExpression->args);
+	if (!IsA(resultIdConst, Const))
+	{
+		/* not sure how to interpret non-const */
+		return;
+	}
+
+	resultIdDatum = resultIdConst->constvalue;
+	resultId = TextDatumGetCString(resultIdDatum);
+
+	resultSize = IntermediateResultSize(resultId);
+	if (resultSize < 0)
+	{
+		/* result does not exist, will probably error out later on */
+		return;
+	}
+
+	resultFormatConst = (Const *) lsecond(funcExpression->args);
+	if (!IsA(resultFormatConst, Const))
+	{
+		/* not sure how to interpret non-const */
+		return;
+	}
+
+	resultFormatDatum = resultFormatConst->constvalue;
+	resultFormatId = DatumGetObjectId(resultFormatDatum);
+
+	if (resultFormatId == BinaryCopyFormatId())
+	{
+		binaryFormat = true;
+
+		/* subtract 11-byte signature + 8 byte header + 2-byte footer */
+		resultSize -= 21;
+	}
+
+	/* start with the cost of evaluating quals */
+	rowCost += relOptInfo->baserestrictcost.per_tuple;
+
+	/* postgres' estimate for the width of the rows */
+	rowSizeEstimate += reltarget->width;
+
+	/* add 2 bytes for column count (binary) or line separator (text) */
+	rowSizeEstimate += 2;
+
+	foreach(typeCell, rangeTableFunction->funccoltypes)
+	{
+		Oid columnTypeId = lfirst_oid(typeCell);
+		Oid inputFunctionId = InvalidOid;
+		Oid typeIOParam = InvalidOid;
+
+		if (binaryFormat)
+		{
+			getTypeBinaryInputInfo(columnTypeId, &inputFunctionId, &typeIOParam);
+
+			/* binary format: 4 bytes for field size */
+			rowSizeEstimate += 4;
+		}
+		else
+		{
+			getTypeInputInfo(columnTypeId, &inputFunctionId, &typeIOParam);
+
+			/* text format: 1 byte for tab separator */
+			rowSizeEstimate += 1;
+		}
+
+		/* add the cost of parsing a column */
+		rowCost += get_func_cost(inputFunctionId) * cpu_operator_cost;
+	}
+
+	/* estimate the number of rows based on the file size and estimated row size */
+	rowCountEstimate = Max(1, (double) resultSize / rowSizeEstimate);
+
+	/* cost of reading the data */
+	ioCost = seq_page_cost * resultSize / BLCKSZ;
+
+	Assert(pathList != NIL);
+
+	/* tell the planner about the cost and row count of the function */
+	path = (Path *) linitial(pathList);
+	path->rows = rowCountEstimate;
```

Comment:
Normally we should also consider row selectivity here, but since (1) we don't use any statistics here, (2) I think overestimating number of rows is better than underestimating number of rows (wrong nested loop joins are usually very expensive), I would pass on this until we use some kind of statistics.

I'm not sure if we can get some kind of good statistics here. Maybe just open an issue for this and `rowSizeEstimate` for now until we can think of a better way.